stuncore/stunreader.cpp

```diff
@@ -153,9 +153,19 @@ HRESULT CStunMessageReader::ValidateMessageIntegrity(uint8_t* key, size_t keylen
     uint8_t hmaccomputed[c_hmacsize] = {}; // zero-init
     unsigned int hmaclength = c_hmacsize;
 #ifndef __APPLE__
-    HMAC_CTX ctx = {};
+    HMAC_CTX* ctx;
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+    HMAC_CTX ctxData;
+    ctx = &ctxData;
+    HMAC_CTX_init(ctx);
 #else
-    CCHmacContext ctx = {};
+    ctx = HMAC_CTX_new();
```

Comment:
Do this instead:

Make ctx a member variable of this class when `OPENSSL_VERSION_NUMBER` is >= 1.1

Instead of calling `HMAC_CTX_new()` in `ValidateMessageIntegritry`, replace that call as you have it with a call to `HMAC_CTX_reset()`

Do the HMAC_CTX_new call in the constructor of this class.  Do the HMAC_CTX_free call in the destructor.

Rationale: The core code is written such that the server can run in a continuous loop without making a single heap allocation.  (i.e. all the variables and objects are allocated up front).   Having a heap allocation per-packet reduces performance.